[dnf5] Package, Reldep, ReldepList, ... classes

j-mracek opened the pull request.

j-mracek committed:
[dnf5] Create a Package class with first functions

It uses package-private.hpp with inlines, that it will be used by
Query class to use advantages of package without requirement
of allocation of memory for package and creation of a Sack smart
pointer.

j-mracek committed:
[dnf5] Create a dummy Sack class

It is required for the Package class.

j-mracek committed:
Add type test for Ids

It is required to ensure that PackageId::id has identical type with
libsolv Id.

j-mracek committed:
Move Dependency structures from dnf4 directory

This is initial commit before refactoring

j-mracek committed:
[dnf5] Rename and refactor Dependency and DependencyContainer

This is the first part that should provide a better interface.

j-mracek committed:
[dnf5] Move IdQueue from dnf4 with minnor changes

Additionally it rename some function according to a new style and adds
one function.

j-mracek committed:
[dnf5] Add additional methods to Package class

They replace functionality from dnf4

j-mracek committed:
[dnf5] Add checksum getters for Package class

j-mracek committed:
[dnf5] Change operator= of ReldepList

j-mracek committed:
[dnf5] Add method and operators for IdQueue

dmach commented on include/libdnf/rpm/chksum.hpp:
```diff
@@ -0,0 +1,66 @@
+class Package;
+
+/// Class contains checksum and checksum type
+class Chksum
```
Rename to 'Checksum' ?
That would mean also renaming the file, header guards, etc.

j-mracek replied:
Done

dmach commented on include/libdnf/rpm/chksum.hpp:
```diff
@@ -0,0 +1,66 @@
+class Chksum
+{
+public:
+    enum class ChksumType {
```
Rename to Type? -> libdnf::rpm::Checksum::Type

j-mracek replied:
Done

dmach commented on include/libdnf/rpm/chksum.hpp:
```diff
@@ -0,0 +1,66 @@
+    std::string chksum;
+    int type;
+};
+/// Constructor requires checksum in hex and libsolv checksum type
```
Add 2 blank lines.

j-mracek replied:
Done

dmach commented on include/libdnf/rpm/reldep.hpp:
```diff
@@ -0,0 +1,143 @@
+class Reldep
+{
+public:
+    enum class ComparisonType {
```
Rename to Cmp or CmpType?
libdnf::rpm::Reldep::Cmp::EQ looks reasonable to me.

j-mracek replied:
Renamed

dmach commented on include/libdnf/rpm/reldep.hpp:
```diff
@@ -0,0 +1,143 @@
+{
+public:
+    enum class ComparisonType {
+        NONE = 0,
+        GT = 1,
+        EQ = 2,
+        GT_EQ = 3,
+        LT = 4,
+        LT_EQ = 6
```
We could either keep these as sequential values or use a bit mask, see:
https://github.com/rpm-software-management/libdnf/pull/930/files#diff-6377fb8e92bc3165e82f0b11978b18dfR32

Also rename `GT_EQ` to `GTE` and `LT_EQ` to `LTE` for consistency with the PR above.

j-mracek replied:
We can do that but I used identical values to libsolv in `pool.h`. See also`static_assert(static_cast<int>(Reldep::ComparisonType::EQ) == REL_EQ,
                  "Reldep::ComparisonType::EQ is not identical to solv/REL_EQ");` in `reldep.cpp`.

dmach replied:
For me, libdnf internal consistency is more important than consistency between libdnf and libsolv.

j-mracek replied:
The approach that I used will not require transformation of libdnf flag to libsolv flag.

j-mracek replied:
I tried to refactor it according to request.

dmach commented on include/libdnf/rpm/chksum.hpp:
```diff
@@ -0,0 +1,66 @@
+    Chksum(const char * chksum, int type);
+
+    std::string chksum;
+    int type;
```
Replace int with the enum class? (the same in the constructor arguments)

dmach commented on include/libdnf/rpm/package.hpp:
```diff
@@ -0,0 +1,287 @@
+
+    /// @replaces dnf:dnf/package.py:attribute:Package.chksum
+    /// @replaces libdnf:libdnf/hy-package-private.hpp:function:dnf_package_get_chksum(DnfPackage *pkg, int *type)
+    Chksum get_chksum();
```
Rename to `get_checksum`

j-mracek replied:
Renamed

dmach commented on include/libdnf/rpm/package.hpp:
```diff
@@ -0,0 +1,287 @@
+
+    /// @replaces dnf:dnf/package.py:attribute:Package.hdr_chksum
+    /// @replaces libdnf:libdnf/hy-package-private.hpp:function:dnf_package_get_hdr_chksum(DnfPackage *pkg, int *type)
+    Chksum get_hdr_chksum();
```
Rename to `get_hdr_checksum`

j-mracek replied:
Renamed

dmach commented on include/libdnf/rpm/package.hpp:
```diff
@@ -0,0 +1,287 @@
+
+    /// @replaces dnf:dnf/package.py:attribute:Package.medianr
+    /// @replaces libdnf:libdnf/hy-package.h:function:dnf_package_get_medianr(DnfPackage * pkg)
+    unsigned long long  get_medianr() noexcept;
```
Name of this attribute is not quite obvious.
I'd consider renaming it to `get_media_number` or at least add a comment explaining what it contains.

j-mracek replied:
Renamed

dmach commented on include/libdnf/rpm/package.hpp:
```diff
@@ -0,0 +1,287 @@
+    /// TODO is_local
+    /// @replaces dnf:dnf/package.py:method:Package.localPkg(self)
+    /// @replaces libdnf:libdnf/dnf-package.h:function:dnf_package_is_local(DnfPackage * pkg)
+    //bool is_local() const;
```
local vs @Commandline repo? Maybe the relation could be explicitly documented.

j-mracek replied:
Local is not only commandline repository, but also packages from local repository.

dmach commented on include/libdnf/rpm/package.hpp:
```diff
@@ -0,0 +1,287 @@
+
+    /// @replaces dnf:dnf/package.py:attribute:Package.medianr
+    /// @replaces libdnf:libdnf/hy-package.h:function:dnf_package_get_medianr(DnfPackage * pkg)
+    unsigned long long  get_rpmdbid() noexcept;
```
Explain what the value represents.

j-mracek replied:
Added an description from dnf doc

dmach commented on include/libdnf/rpm/reldep.hpp:
```diff
@@ -0,0 +1,143 @@
+/// @replaces libdnf/dnf-reldep.h:struct:DnfReldep
+/// @replaces libdnf/repo/solvable/Dependency.hpp:struct:Dependency
+/// @replaces hawkey:hawkey/__init__.py:class:Reldep
+class Reldep
```
Maybe add a comment explaining that this is a Relational dependency

j-mracek replied:
Added

dmach commented on libdnf/rpm/reldep.cpp:
```diff
@@ -0,0 +1,108 @@
+ReldepId
+Reldep::get_reldep_id(Sack * sack, const std::string & reldep_str)
+{
+    if (reldep_str[0] == '(') {
```
Shouldn't we move this to a public method `bool is_rich_dependency()`?

j-mracek replied:
In case that the test will be required on multiple sides, then yes.

dmach commented on libdnf/rpm/reldep_splitter.cpp:
```diff
@@ -0,0 +1,90 @@
+{
+    enum { NAME = 1, CMP_TYPE = 2, EVR = 3};
+    std::smatch match;
+    if (!std::regex_match(reldep_str, match, RELDEP_REGEX)) {
```
any chance to use libsolv's or rpm's internal functions instead?

j-mracek replied:
So far I did not find anything in libsolv.

dmach replied:
don't worry about it then; resolving this conversation

dmach commented on libdnf/rpm/solv/id_queue.hpp:
```diff
@@ -0,0 +1,119 @@
+    Id operator[](int index) const;
+
+    void push_back(Id id);
+    void push_back(Id id1, Id id2);
```
`push_back(Id id)` is obvious, but why there's the 2nd version with 2 Ids? Could you document what it does?

j-mracek replied:
It is used for a libsolv job. I will add a note.

dmach commented on libdnf/rpm/solv/id_queue.hpp:
```diff
@@ -0,0 +1,119 @@
+inline void IdQueue::clear() noexcept { queue_empty(&queue); }
+
+inline void
+IdQueue::insert(const IdQueue & src)
```
What does the insert do? I'd expect it inserting an Id at a specified position.

j-mracek replied:
It inserts the new data at the end position. I can rename it, or add an argument, or overloaded, or add description. What do you think?

dmach replied:
Based on the argument being another an IdQueue containing multiple values, I expect it extends the original queue.
I'd use `extend` in that case.

j-mracek replied:
Renamed to append()

j-mracek committed:
[dnf5] Make reldep parser faster

It also moved and renamed ReldepSplitter.

Additionally it fixes the problem when only name reldem is used after
parsing of reldep with evr and comperator.

j-mracek committed:
Add unittest for ReldepParser

j-mracek committed:
Add tests for IdQueue

j-mracek commented on include/libdnf/rpm/checksum.hpp:
```diff
@@ -0,0 +1,70 @@
+    Checksum(const char * checksum, int libsolv_type);
+
+    std::string checksum;
+    int libsolv_type;
```
Originally from @dmach:
`Replace int with the enum class? (the same in the constructor arguments)`

I decide to store Id/int from libsolv, because it requires only one conversion to enum (`get_type()`). In case of storage of enum value it requires conversion for constructor and when `get_type_str()` is called. `get_type_str()` uses convertor implemented in libsolv. Additional advantage is that when libsolv will start to use additional checksum types, libdnf type will tell 'inknown', but string will correctly report the new type.

j-mracek commented on libdnf/rpm/solv/id_queue.hpp:
```diff
@@ -0,0 +1,137 @@
+    Queue & get_queue() noexcept;
+    int size() const noexcept;
+    void clear() noexcept;
+    void append(const IdQueue & src);
```
Added from outdated:
@dmach
dmach yesterday Member

What does the insert do? I'd expect it inserting an Id at a specified position.
@j-mracek
j-mracek yesterday Author Member

It inserts the new data at the end position. I can rename it, or add an argument, or overloaded, or add description. What do you think?
@dmach
dmach 10 hours ago Member

Based on the argument being another an IdQueue containing multiple values, I expect it extends the original queue.
I'd use extend in that case.

j-mracek replied:
After discussion with jrohel and mblaha, the function insert was renamed to append.

j-mracek committed:
[dnf5] Rename Reldep::ComparisonType to CmpType

j-mracek committed:
fixup! [dnf5] Make reldep parser faster

j-mracek committed:
[dnf5] Add changes according to clang recommendation

jrohel commented on include/libdnf/rpm/checksum.hpp:
```diff
@@ -0,0 +1,70 @@
+
+    /// Return checksum in hex format
+    const std::string & get_checksum() const noexcept { return checksum; };
+    
```
Remove spaces from empty line.

jrohel commented on include/libdnf/rpm/reldep.hpp:
```diff
@@ -0,0 +1,144 @@
+
+    /// @replaces libdnf/repo/solvable/Dependency.hpp:method:Dependency(const Dependency & dependency);
+    Reldep(const Reldep & reldep);
+    
```
Remove spaces from empty line.

jrohel commented on libdnf/rpm/package.cpp:
```diff
@@ -0,0 +1,402 @@
+Package::get_name_cstring() const noexcept
+{
+    Pool * pool = sack->pImpl->pool;
+    return libdnf::rpm::solv::get_name(pool, id);
```
The code is in the `libdnf::rpm` namespace. Please use `solv::get_name(pool, id)`. The same is repeated in next methods.

Second possibility is using absolute namespace path. Starts with `::`. Example: `::libdnf::rpm::solv::get_name(pool, id)`. I do not recommend absolute path in these cases.

Using fullpath without `::` is possible but it is considered as relative path. If the relative path does not exist in current namespace, then compiler looks for the same relative path in one level up. And it can repeat up to the global namespace. So, it is working. But I think, the shortest method `solv::get_name(pool, id)` is what we want in this case.

j-mracek replied:
Replaced

jrohel commented on libdnf/rpm/reldep.cpp:
```diff
@@ -0,0 +1,108 @@
+#include "sack_impl.hpp"
+#include "solv/reldep_parser.hpp"
+
+/* workaround, libsolv lacks 'extern "C"' in its header file */
```
Style: Do not use multi line comments in code. It is on more places.

j-mracek replied:
Fixed

jrohel commented on libdnf/rpm/reldep.cpp:
```diff
@@ -0,0 +1,108 @@
+const char * Reldep::get_version() const { return pool_id2evr(sack->pImpl->pool, id.id); }
+const char * Reldep::to_string() const { return pool_dep2str(sack->pImpl->pool, id.id); }
+
+ReldepId
```
Style: Put the return type on the same line as method name. Try clang-format. The same repeats on lot of places.

j-mracek replied:
Fixed by clang-format

jrohel commented on libdnf/rpm/reldep.cpp:
```diff
@@ -0,0 +1,108 @@
+        Pool * pool = sack->pImpl->pool;
+        Id id = pool_parserpmrichdep(pool, reldep_str.c_str());
+        if (!id)
+            throw std::runtime_error("Cannot parse a dependency string");
```
Use the libdnf specific exception class. `RuntimeError()`. Or much better define exception class for this case.
But it can be fixed later if you add `TODO:`

j-mracek replied:
TODO was added. Still I am not sure if trow of an exception is a right approach.

jrohel commented on include/libdnf/rpm/reldep.hpp:
```diff
@@ -0,0 +1,144 @@
+
+    /// @replaces libdnf/repo/solvable/Dependency.hpp:method:toString()
+    /// @replaces libdnf/dnf-reldep.h:function:dnf_reldep_to_string(DnfReldep *reldep)
+    const char * to_string() const;
```
I looked into libsolv source code. In some cases the return value is allocated in temporary space. And in the libsolv header is: "might alloc tmpspace". Please add this information to the method description.

j-mracek replied:
const char * replaced by std::string.

jrohel commented on libdnf/rpm/reldep_list.cpp:
```diff
@@ -0,0 +1,139 @@
+        }
+    }
+
+    return pImpl->sack->pImpl->pool == other.pImpl->sack->pImpl->pool;
```
OK. The code compare two queues. And after for cycle pass it is found that the `pools` are different. Is the order correct/optimal?

j-mracek replied:
I am also not sure

jrohel commented on libdnf/rpm/reldep_list.cpp:
```diff
@@ -0,0 +1,139 @@
+        add(id);
+        return true;
+    }
+    catch (...) {
```
Using libdnf specific exceptions and catch only they is better. Probably we do not want silently catch all errors eg. "memory overflow".
Add `TODO:`

j-mracek replied:
Added TODO:

jrohel commented on libdnf/rpm/solv/package-private.hpp:
```diff
@@ -0,0 +1,451 @@
+
+namespace libdnf::rpm::solv {
+
+const char * BASE_EPOCH = "0";
```
It is variable that points to const C string.
Probably you want to declare constant `constexpr const char * BASE_EPOCH = "0";`

j-mracek replied:
Added `constexpr`

jrohel commented on libdnf/rpm/solv/reldep_parser.hpp:
```diff
@@ -0,0 +1,79 @@
+
+inline const char * ReldepParser::get_name_cstr() const noexcept
+{
+    return name.empty() ? NULL : name.c_str();
```
`NULL` -> `nullptr`

j-mracek replied:
Fixed

jrohel commented on libdnf/rpm/solv/reldep_parser.hpp:
```diff
@@ -0,0 +1,79 @@
+
+inline const char * ReldepParser::get_evr_cstr() const noexcept
+{
+    return evr.empty() ? NULL : evr.c_str();
```
`NULL` -> `nullptr`

j-mracek replied:
Fixed

jrohel commented on libdnf/rpm/solv/package-private.hpp:
```diff
@@ -0,0 +1,451 @@
+    dataiterator_init(&di, pool, solvable->repo, package_id.id, SOLVABLE_FILELIST, nullptr,
+                      SEARCH_FILES | SEARCH_COMPLETE_FILELIST);
+    while (dataiterator_step(&di) != 0) {
+       ret.push_back(di.kv.str);
```
`ret.emplace_back(di.kv.str);`

j-mracek replied:
fixed

j-mracek committed:
[dnf5] Change return value of to_string method

The original code returned const char * that was not permanent,
therefore to make a copy and return std::string is more secure way.
